### PR TITLE
Add aten.bucketize op and its decomposition

### PR DESCRIPTION
**Pull request by li-plus**
This PR supported aten.bucketize op by decomposition. It can be lowered to LINALG and MHLO backends. Below is the equivalent python code for decomposition.

```python
def aten_bucketize(input, boundaries, out_int32, right):
    unsqz_input = input.unsqueeze(-1)
    if not right:
        comparison = unsqz_input <= boundaries
    else:
        comparison = unsqz_input < boundaries
    indices = torch.argmax(comparison.float(), dim=-1)
    within_bound = comparison[..., -1]
    result = torch.where(within_bound, indices, boundaries.shape[0])
    if out_int32:
        result = result.int()
    return result
```

To make this decomposition work, this PR also supported lowering `AtenLeTensorOp` and `AtenGeTensorOp` to LINALG and added e2e tests for them.

### CODE REVIEW

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 1]
```diff
@@ -606,6 +637,128 @@ class DecomposeAtenArgMaxOp : public OpRewritePattern<AtenArgmaxOp> {
...
+//         comparison = unsqz_input <= boundaries
+//     else:
+//         comparison = unsqz_input < boundaries
+//     indices = torch.argmax(comparison.float(), dim=-1)
+//     within_bound = comparison[..., -1]
+//     result = torch.where(within_bound, indices, boundaries.shape[0])
+//     if out_int32:
+//         result = result.int()
+//     return result
+//
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:654]: Thanks for this! It helped a lot

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 2]
```diff
@@ -3786,6 +3923,7 @@ class DecomposeComplexOpsPass
     addPatternIfTargetOpIsIllegal<DecomposeAtenLeakyReluOp>(patterns);
     addPatternIfTargetOpIsIllegal<DecomposeAtenLeakyReluBackwardOp>(patterns);
     addPatternIfTargetOpIsIllegal<DecomposeAtenNewEmptyStridedOp>(patterns);
+    addPatternIfTargetOpIsIllegal<DecomposeAtenBucketizeTensorOp>(patterns);
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:3926]: You need to also add the op here: https://github.com/llvm/torch-mlir/blob/95361747c2c350886d192261667bfc57d870971d/lib/Dialect/Torch/Transforms/LowerToBackendContract.cpp#L335

File: python/torch_mlir/csrc/base_lazy_backend/shape_inference.cpp [hunk 3]
```diff
@@ -49,5 +49,12 @@ std::vector<torch::lazy::Shape> compute_shape_where(
   return {Shape(self.scalar_type(), self.sizes().vec())};
 }
 
+std::vector<torch::lazy::Shape> compute_shape_bucketize(
```
Comment (ramiro050) [inline python/torch_mlir/csrc/base_lazy_backend/shape_inference.cpp:52]: @antoniojkim, can you review the changes in this file?

**Review by ramiro050 — CHANGES_REQUESTED** [review]
Great work!

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp [hunk 4]
```diff
@@ -117,6 +134,46 @@ static Value createCalculationForMathOpWithDtypeConversion(
   return b.create<MathOpTy>(loc, arg);
 }
 
+template <typename OpTy>
+static Value createCompareTensorOp(OpBuilder &b, Location loc, OpTy op,
+                                   Value lhs, Value rhs) {
+  static_assert(std::is_same<OpTy, AtenLtTensorOp>() ||
+                    std::is_same<OpTy, AtenLeTensorOp>() ||
```
[the remark this answers is not in the document]
Comment (li-plus) [reply lib/Conversion/TorchToLinalg/Uncategorized.cpp:141]: This is exactly what clang-format produces. Kind of counter-intuitive.

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp [hunk 5]
```diff
@@ -117,6 +134,46 @@ static Value createCalculationForMathOpWithDtypeConversion(
...
+  // TODO: Type promotion in case of different `lhsDtype` and `rhsDtype` needs
+  // to be handled.
+  if (lhsDtype != rhsDtype) {
+    op.emitError("unimplemented: lhs and rhs dtype must be same");
+    return nullptr;
+  }
+
+  Type elementalType =
+      op.getSelf().getType().template cast<BaseTensorType>().getDtype();
+  if (std::is_same<OpTy, AtenLtTensorOp>()) {
```
[the remark this answers is not in the document]
Comment (li-plus) [reply lib/Conversion/TorchToLinalg/Uncategorized.cpp]: Done!

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 6]
```diff
@@ -171,6 +171,37 @@ static Value createSoftmaxBackwardCommonKernel(PatternRewriter &rewriter,
...
+  if (!inputType.hasSizes()) {
+    return rewriter.notifyMatchFailure(op, "input tensor must have size");
+  }
+
+  SmallVector<int64_t> unsqueezedShape;
+  ArrayRef<int64_t> inputShape = inputType.getSizes();
+  int64_t dimInt = 0;
+  if (matchPattern(dim, m_TorchConstantInt(&dimInt))) {
+    // `input` has a reduced rank. Hence add 1.
+    int64_t unsqueezedRank = inputShape.size() + 1;
```
[the remark this answers is not in the document]
Comment (li-plus) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Good catch! Done.

**Comment by ramiro050** [comment]
@li-plus, I've added you as a collaborator in torch-mlir so that you can have write access. You should've received an email about it

File: python/torch_mlir/csrc/base_lazy_backend/shape_inference.cpp (hunk 3 above)
Comment (antoniojkim) [reply python/torch_mlir/csrc/base_lazy_backend/shape_inference.cpp:52]: changes in this file look good to me

**Review by ramiro050 — APPROVED** [review]

**Comment by li-plus** [comment]
> @li-plus, I've added you as a collaborator in torch-mlir so that you can have write access. You should've received an email about it

Thanks! It really helps a lot.

**Pull request merged**